core/services/aws/donorSearch/queryEligibleDonors.ts

```diff
@@ -0,0 +1,280 @@
+  if (geohashesToProcess.length === 0 ||
+    processedGeohashCount >= MAX_GEOHASHES_PER_PROCESSING_BATCH ||
+    Object.keys(eligibleDonors).length >= totalDonorsToNotify) {
+    return { newEligibleDonors: eligibleDonors, processedGeohashCount }
+  }
+
+  const geohashToProcess = geohashesToProcess[0]
+  const geohashCachePrefix = geohashToProcess.slice(0, CACHE_GEOHASH_PREFIX_LENGTH)
+  const cacheKey = `${city}-${requestedBloodGroup}-${geohashCachePrefix}`
+  const cachedGroupedGeohash = geohashCache.get(cacheKey)
+
+  if (cachedGroupedGeohash === undefined) {
+    const queriedDonors = await donorSearchService.queryGeohash(
+      city,
+      requestedBloodGroup,
+      geohashCachePrefix,
+      new DynamoDbTableOperations<LocationDTO, LocationFields, LocationModel>(
+        new LocationModel()
+      )
+    )
+    updateGroupedGeohashCache(queriedDonors, cacheKey)
+  }
+
+  const cachedDonorMap = geohashCache.get(cacheKey) as GeohashDonorMap
+  const donors = cachedDonorMap[geohashToProcess] ?? []
+
+  const newDonors = donors.reduce<Record<string, EligibleDonorInfo>>(
+    (donorAccumulator, donor) => {
+      const donorDistance = getDistanceBetweenGeohashes(seekerGeohash, geohashToProcess)
+      if (
+        donor.userId !== seekerId &&
+        (donorAccumulator[donor.userId] === undefined ||
+          donorAccumulator[donor.userId].distance < donorDistance)
+      ) {
+        donorAccumulator[donor.userId] = {
+          locationId: donor.locationId,
+          distance: donorDistance
+        }
+      }
+      return donorAccumulator
+    },
+    { ...eligibleDonors }
+  )
+
+  return getNewDonorsInNeighborGeohash(
+    geohashesToProcess.slice(1), city,
+    requestedBloodGroup,
+    seekerGeohash,
+    seekerId,
+    newDonors,
+    totalDonorsToNotify,
+    processedGeohashCount + 1
+  )
+}
+
+async function getClosestDonorsAcrossAll(
+  city: string,
+  requestedBloodGroup: string,
+  seekerGeohash: string,
+  seekerId: string,
+  totalDonorsToNotify: number
+): Promise<Record<string, EligibleDonorInfo>> {
+  const allDonors = await donorSearchService.queryGeohash(
+    city,
+    requestedBloodGroup,
+    '',
+    new DynamoDbTableOperations<LocationDTO, LocationFields, LocationModel>(new LocationModel())
+  )
+  const newFoundEligibleDonors = allDonors.reduce<Record<string, EligibleDonorInfo>>(
+    (accumulator, donor) => {
+      const donorDistance = getDistanceBetweenGeohashes(seekerGeohash, donor.geohash)
+
+      if (
+        donor.userId !== seekerId &&
+        (accumulator[donor.userId] === undefined ||
+          accumulator[donor.userId].distance > donorDistance)
+      ) {
+        accumulator[donor.userId] = {
+          locationId: donor.locationId,
+          distance: donorDistance
+        }
+      }
+      return accumulator
+    },
+    {}
+  )
+  const closestDonors = Object.entries(newFoundEligibleDonors)
```

Comment:
you meant donor donorsSortedByDistance?
I feel like the original reduce
`newFoundEligibleDonors` could have done this as well. It could do an insertion sort. 
I am also not sure how the sort stays in place cause you're using reduce and then using it to create an object, javascript objects are sorted on insertion and it uses a lexical order, which could work for distance too.

Reply:
We are not using the anymore